parser.js

```diff
@@ -124,56 +135,90 @@ function generateAST(code, options) {
     warnedAboutTSVersion = true;
   }
 
-  // Even if jsx option is set in typescript compiler, filename still has to
-  // contain .tsx file extension
-  const FILENAME = extra.ecmaFeatures.jsx ? 'estree.tsx' : 'estree.ts';
-
-  const compilerHost = {
-    fileExists() {
-      return true;
-    },
-    getCanonicalFileName() {
-      return FILENAME;
-    },
-    getCurrentDirectory() {
-      return '';
-    },
-    getDefaultLibFileName() {
-      return 'lib.d.ts';
-    },
-
-    // TODO: Support Windows CRLF
-    getNewLine() {
-      return '\n';
-    },
-    getSourceFile(filename) {
-      return ts.createSourceFile(filename, code, ts.ScriptTarget.Latest, true);
-    },
-    readFile() {
-      return null;
-    },
-    useCaseSensitiveFileNames() {
-      return true;
-    },
-    writeFile() {
-      return null;
-    }
-  };
+  let relevantProgram = undefined;
+  let ast = undefined;
+  const shouldProvideParserServices = extra.project && extra.project.length > 0;
 
-  const program = ts.createProgram(
-    [FILENAME],
-    {
-      noResolve: true,
-      target: ts.ScriptTarget.Latest,
-      jsx: extra.ecmaFeatures.jsx ? 'preserve' : undefined
-    },
-    compilerHost
-  );
+  if (shouldProvideParserServices) {
+    const FILENAME = options.filePath;
+    const programs = calculateProjectParserOptions(code, options, extra);
+    for (const program of programs) {
+      ast = program.getSourceFile(FILENAME);
 
-  const ast = program.getSourceFile(FILENAME);
+      if (ast !== undefined) {
+        relevantProgram = program;
+        break;
+      }
+    }
+  }
+
+  if (ast === undefined) {
```

Comment:
This function is getting pretty long so I think it could be broken down:

```ts
  const { ast, relevantProgram } = getProgramAndAst();
  ...
}

function getProgramAndAst...() {
  return extra.project && extra.project.length > 0 && getFromProject(...) || createNewProject(...);
}
function getFromProject(...) {
  return firstDefined(calculateProjectParserOptions(code, options, extra), relevantProgram => {
      const ast = relevantProgram.getSourceFile(options.filePath);
      return ast && { ast, relevantProgram };
  });
}
function createNewProject(...) {
   ... all the code here, and `return { ast, relevantProgram };` at the end.
}
```

Reply:
@uniqueiniquity When the code is organized like this, it looks like without special configuration it will call `createNewProject` again every time.

Reply:
Ah, I see your point. That's actually not a regression, but rather the existing behavior. I'd be happy to suggest they change that, but I'd rather that be addressed separately so my contribution is wholly enabled or disabled by the `project` parser option.

Reply:
Refactored as suggested